meeteval/wer/matching/cy_levenshtein.pyx

```diff
@@ -0,0 +1,279 @@
+            vector[uint] reference,
+            vector[uint] hypothesis,
+            vector[pair[T, T]] reference_timing,
+            vector[pair[T, T]] hypothesis_timing,
+            uint cost_del,
+            uint cost_ins,
+            uint cost_sub,
+            uint cost_cor,
+    )
+
+    struct LevenshteinStatistics:
+        uint insertions
+        uint deletions
+        uint substitutions
+        uint correct
+        uint total
+        vector[pair[uint, uint]] alignment
+
+    LevenshteinStatistics time_constrained_levenshtein_distance_with_alignment_[T](
+            vector[uint] reference,
+            vector[uint] hypothesis,
+            vector[pair[T, T]] reference_timing,
+            vector[pair[T, T]] hypothesis_timing,
+            uint cost_del,
+            uint cost_ins,
+            uint cost_sub,
+            uint cost_cor,
+            uint eps
+    )
+
+
+def obj2vec(a, b):
+    int2sym = dict(enumerate(sorted(set(a) | set(b))))
+    sym2int = {v: k for k, v in int2sym.items()}
+    return [sym2int[a_] for a_ in a], [sym2int[b_] for b_ in b]
+
+def levenshtein_distance(
+        reference,
+        hypothesis,
+):
+    reference, hypothesis = obj2vec(reference, hypothesis)
+
+    return levenshtein_distance_(reference, hypothesis)
+
+def levenshtein_distance_custom_cost(
+        reference,
+        hypothesis,
+        cost_del,
+        cost_ins,
+        cost_sub,
+        cost_cor,
+):
+    reference, hypothesis = obj2vec(reference, hypothesis)
+
+    return levenshtein_distance_custom_cost_(
+        reference, hypothesis, cost_del, cost_ins, cost_sub, cost_cor
+    )
+
+def _validate_inputs(reference, hypothesis, reference_timing, hypothesis_timing):
+    if len(reference) != len(reference_timing):
+        raise ValueError(
+            f'reference and reference_timing have mismatching lengths '
+            f'{len(reference)} != {len(reference_timing)}'
+        )
+    if len(hypothesis) != len(hypothesis_timing):
+        raise ValueError(
+            f'hypothesis and hypothesis_timing have mismatching lengths '
+            f'{len(hypothesis)} != {len(hypothesis_timing)}'
+        )
+
+    reference_timing = np.array(reference_timing)
+    hypothesis_timing = np.array(hypothesis_timing)
+
+    assert len(reference) == 0 or reference_timing.shape == (len(reference), 2), (reference_timing.shape, len(reference))
+    assert len(hypothesis) == 0 or hypothesis_timing.shape == (len(hypothesis), 2), (hypothesis_timing.shape, len(hypothesis))
+    assert len(reference) == 0 or len(hypothesis) == 0 or reference_timing.dtype == hypothesis_timing.dtype, (reference_timing.dtype, hypothesis_timing.dtype)
+
+    def check(timing):
+        # end >= start
+        if np.any(timing[:, 1] < timing[:, 0]):
+            raise ValueError(f'The end time of an interval must not be smaller than its begin time')
+        # start values are increasing
+        if np.any(np.diff(timing[:, 0]) < 0):
```

Comment:
Does the code work correctly, when the "next" end_time is smaller?

Reply:
Yes, I now keep track of the largest seen end value for the pruning optimization. Only the start values have to be sorted